File: src/p4.py
import re

file1 = open("./data/d3.txt")
lines = file1.readlines()

def check(min, max, letter, password):
    min = int(min)
    max = int(max)

    a = password[min - 1]
    b = password[max - 1]

    if a != b and (a == letter or b == letter):
        return 1
    else:
        return 0


def testLine(x):
    x = re.match(r"(\d+)-(\d+) (\w): (\w+)", x)
    return check(x.group(1), x.group(2), x.group(3), x.group(4))


print(sum(testLine(line) for line in lines))

File: src/p4.rs
use std::lazy::Lazy;

use regex::Regex;

#[derive(Debug)]
struct PWTest<'a> {
    min: usize,
    max: usize,
    letter: char,
    text: &'a str,
}

impl<'a> PWTest<'a> {
    fn validate(&self) -> usize {
        let a = self.text.chars().nth(self.min - 1).unwrap();
        let b = self.text.chars().nth(self.max - 1).unwrap();

        if a != b && (a == self.letter || b == self.letter) {
            1
        } else {
            0
        }
    }
}

const REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"(\d+)-(\d+) (\w): (\d+)").unwrap());

impl<'a> PWTest<'a> {
    fn new(s: &'a str) -> Self {
        let m: Vec<_> = REGEX
            .captures(&s)
            .unwrap()
            .iter()
            .map(|x| x.unwrap().as_str())
            .collect();

        Self {
            min: m[1].parse().unwrap(),
            max: m[2].parse().unwrap(),
            letter: m[3].chars().nth(0).unwrap(),
            text: m[4],
        }
    }
}

pub fn main() {
    let valids: usize = crate::input(3)
        .lines()
        .map(PWTest::new)
        .map(|x| x.validate())
        .sum();
    println!("valid {}", valids);
}
